checks.d/splunk_topology.py

```diff
@@ -0,0 +1,190 @@
+"""
+    StackState.
+    Splunk topology extraction
+"""
+
+# 3rd party
+import requests
+from urllib import quote
+import time
+
+# project
+from checks import AgentCheck, CheckException
+
+
+class SavedSearch:
+    def __init__(self, saved_search):
+        self.name = saved_search['name']
+        self.element_type = saved_search['element_type']
+        self.parameters = saved_search['parameters']
+
+
+class InstanceConfig:
+    def __init__(self, instance, init_config):
+        default_timeout = init_config.get('default_timeout', 5)
+        max_retry_count = init_config.get('max_retry_count', 3)
+        seconds_between_retries = init_config.get('seconds_between_retries', 1)
+        polling_interval = init_config.get('polling_interval', 15)
+
+        self.base_url = instance['url']
+        self.username = instance['username']
+        self.password = instance['password']
+        self.timeout = float(instance.get('timeout', default_timeout))
+        self.max_retry_count = int(instance.get('max_retry_count', max_retry_count))
+        self.seconds_between_retries = int(instance.get('seconds_between_retries', seconds_between_retries))
+        self.polling_interval = int(instance.get('polling_interval', polling_interval))
+
+    def get_auth_tuple(self):
+        return self.username, self.password
+
+
+class Instance:
+    INSTANCE_TYPE = "splunk"
+
+    def __init__(self, instance, init_config):
+        self.instance_config = InstanceConfig(instance, init_config)
+        self.saved_searches = [SavedSearch(saved_search) for saved_search in instance['saved_searches']]
+        self.instance_key = {
+            "type": self.INSTANCE_TYPE,
+            "url": self.instance_config.base_url
+        }
+        self.tags = instance.get('tags', [])
+        self.last_successful_poll = 0
+
+    def should_poll(self, time_seconds):
+        return self.last_successful_poll == 0 or time_seconds >= self.last_successful_poll + self.instance_config.polling_interval
+
+
+class SplunkTopology(AgentCheck):
+    # SERVICE_CHECK_NAME = "splunk.topology_information"
+    # service_check_needed = True
+
+    def __init__(self, name, init_config, agentConfig, instances=None):
+        super(SplunkTopology, self).__init__(name, init_config, agentConfig, instances)
+        # Data to keep over check runs, keyed by instance url
+        self.instance_data = dict()
+
+    def check(self, instance):
+        if 'url' not in instance:
+            raise CheckException('Splunk topology instance missing "url" value.')
+
+        if instance["url"] not in self.instance_data:
+            self.instance_data[instance["url"]] = Instance(instance, self.init_config)
```

Comment:
Ok, I looked at the implementation and the polling interval now determines wether or not the results of a saved search are collected, instead of whether or not the saved search is actually dispatched. This is not what I would expect.

I also don't think its a good idea to set the interval on a saved search, because I can imagine you would want a saved search returning relations and a saved search returning components in the same snapshot. By altering the polling intervals, you would sometimes have things inside the snapshot and sometimes not, causing components to get deleted from stackstate. So therefore I think that if data belongs to the same snapshot, it should also have the same polling interval.

For me it seems like we need per instance another way of grouping saved searches together, e.g. we could define a list of topologies/snapshots inside one instance. A topology/snapshot has then settings like polling_interval etc., and a collection of saved searches for components and relations. Grouping them together then would mean they belong to the same snapshot.

However it may be too much work, and we could stick with a simple version which just puts everything inside one snapshot, and has one polling interval. (Basically what we had before)